File: python/flyql/renderers/registry.py
from typing import Any, Callable, Dict, List, Optional

from .base import Renderer

DiagnoseHook = Callable[[Any, List[Dict[str, Any]]], List[Any]]


class RendererRegistry:
    def __init__(self) -> None:
        self._renderers: Dict[str, Renderer] = {}
        self._diagnose: Optional[DiagnoseHook] = None

    def get(self, name: str) -> Optional[Renderer]:
        return self._renderers.get(name)

    def register(self, renderer: Renderer) -> None:
        if renderer.name in self._renderers:
            raise ValueError(f"Renderer '{renderer.name}' is already registered")
        self._renderers[renderer.name] = renderer

    def names(self) -> List[str]:
        return list(self._renderers.keys())

    def set_diagnose(self, fn: DiagnoseHook) -> None:
        self._diagnose = fn

    def get_diagnose(self) -> Optional[DiagnoseHook]:
        return self._diagnose


def default_registry() -> RendererRegistry:
    return RendererRegistry()

File: javascript/packages/flyql/src/renderers/registry.js
export class RendererRegistry {
    constructor() {
        this._renderers = new Map()
        this._diagnose = null
    }

    get(name) {
        return this._renderers.get(name) || null
    }

    register(renderer) {
        if (this._renderers.has(renderer.name)) {
            throw new Error(`Renderer '${renderer.name}' is already registered`)
        }
        this._renderers.set(renderer.name, renderer)
    }

    names() {
        return Array.from(this._renderers.keys())
    }

    setDiagnose(fn) {
        this._diagnose = fn
    }

    getDiagnose() {
        return this._diagnose
    }
}

export function defaultRegistry() {
    return new RendererRegistry()
}
